core/embed/trezorhal/stm32f4/systick.c

```diff
@@ -19,52 +17,193 @@
 
-#include "systemview.h"
+// SysTick driver state
+typedef struct {
+  // Set if the driver is initialized
+  bool initialized;
+  // Number of hw cycles per millisecond (tick period)
+  uint32_t cycles_per_ms;
+  // Number of hw cycles per microsecond
+  uint32_t cycles_per_us;
+  // Current tick value in hardware cycles
+  volatile uint64_t cycles;
+  // Number of ticks (ms) since the system start
+  volatile uint32_t ticks;
+} systick_driver_t;
+
+static systick_driver_t g_systick_driver = {
+    .initialized = false,
+};
+
+void systick_init(void) {
+  systick_driver_t* drv = &g_systick_driver;
+
+  if (drv->initialized) {
+    return;
+  }
+
+  drv->cycles = 0;
+  drv->ticks = 0;
 
-extern __IO uint32_t uwTick;
+  // Set 1ms tick period
+  drv->cycles_per_ms = HAL_RCC_GetSysClockFreq() / 1000;
+  drv->cycles_per_us = drv->cycles_per_ms / 1000;
+
+  // Initialize and enable SysTick timer
+  SysTick_Config(drv->cycles_per_ms);
+
+  // We need to ensure that SysTick has the expected priority.
+  // The SysTick priority is configured in the boardloader,
+  // and some early versions didn't set this properly.
+  NVIC_SetPriority(SysTick_IRQn, IRQ_PRI_NORMAL);
+
+  drv->initialized = true;
+}
+
+void systick_deinit(void) {
+  systick_driver_t* drv = &g_systick_driver;
+
+  if (!drv->initialized) {
+    return;
+  }
+
+  NVIC_DisableIRQ(SysTick_IRQn);
+  SysTick->CTRL = 0;
+  NVIC_ClearPendingIRQ(SysTick_IRQn);
+
+  drv->initialized = false;
+}
+
+void systick_update_freq(void) {
+  systick_driver_t* drv = &g_systick_driver;
+
+  if (!drv->initialized) {
+    return;
+  }
+
+  uint32_t clock_freq = HAL_RCC_GetSysClockFreq();
+  drv->cycles_per_ms = clock_freq / 1000;
+  drv->cycles_per_us = drv->cycles_per_ms / 1000;
+  SysTick_Config(drv->cycles_per_ms);
+
+  // We need to ensure that SysTick has the expected priority.
+  // The SysTick priority is configured in the boardloader,
+  // and some early versions didn't set this properly.
+  NVIC_SetPriority(SysTick_IRQn, IRQ_PRI_NORMAL);
+}
+
+uint64_t systick_cycles(void) {
+  systick_driver_t* drv = &g_systick_driver;
+
+  irq_key_t irq_key = irq_lock();
+
+  // Current value of the SysTick counter
+  uint32_t val = SysTick->VAL;
+
+  // Check if the SysTick has already counted down to 0 or wrapped around
+  // Reading CTRL register automatically clears COUNTFLAG
+  if ((SysTick->CTRL & SysTick_CTRL_COUNTFLAG_Msk) != 0) {
+    // Re-read the current value of the SysTick counter
+    val = SysTick->VAL;
+    // Update the hardware cycles counter
+    // Since we have cleared COUNTFLAG, SysTick_Handler will not increment
+    drv->cycles += drv->cycles_per_ms;
+    // Increment regular ticks counter
+    drv->ticks++;
+  }
+
+  uint64_t cycles = drv->cycles + ((val > 0) ? (drv->cycles_per_ms - val) : 0);
+
+  irq_unlock(irq_key);
+
+  return cycles;
+}
+
+uint64_t systick_us_to_cycles(uint64_t us) {
+  systick_driver_t* drv = &g_systick_driver;
+
+  return us * drv->cycles_per_us;
```

Comment:
The idea is as follows:

1) After a restart, until you call `systick_init()`, these functions return zero (since global variable is set to 0)
2) Once you call `systick_init()`, they return the proper incrementing value.  
3) After you call `systick_deinit()`, they return the last value (they don't reset to zero).

In cases 1 and 3, nobody should call the systick functions, but if they do, they will get expected values.

This behavior was consistent with the emu implementation until we had to implement lazy initialization due to unit tests. However, I hope we will fix it in the future.

Reply:
I agree with the proposed behavior.

The unix impl has this:
```
uint32_t systick_ms() {
  systick_driver_t* drv = &g_systick_driver;

  if (!drv->initialized) {
    systick_init();  // temporary fix fot rust tests
    // return 0;
  }
```

so before the temporary fix, it would return zero in both 1) and 3), therefore the behavior was not consistent. Which is moot point for now, but when we can, it would be preferable to fix it.


In F4 implementation of `systick_us` there is division by zero if called before initialization.